src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs

```diff
@@ -24,25 +30,48 @@ private static unsafe string GenerateId(long id)
             // and ~600% faster than calling long.ToString() on x86 in tight loops of 1 million+ iterations
             // See: https://github.com/aspnet/Hosting/pull/385
 
+#if NETCOREAPP2_1
+            return string.Create(13, id, _spanAction);
+#else
             // stackalloc to allocate array on stack rather than heap
             char* charBuffer = stackalloc char[13];
+            ref char encodeChars = ref _encode32Chars[0];
 
-            charBuffer[0] = _encode32Chars[(int)(id >> 60) & 31];
-            charBuffer[1] = _encode32Chars[(int)(id >> 55) & 31];
-            charBuffer[2] = _encode32Chars[(int)(id >> 50) & 31];
-            charBuffer[3] = _encode32Chars[(int)(id >> 45) & 31];
-            charBuffer[4] = _encode32Chars[(int)(id >> 40) & 31];
-            charBuffer[5] = _encode32Chars[(int)(id >> 35) & 31];
-            charBuffer[6] = _encode32Chars[(int)(id >> 30) & 31];
-            charBuffer[7] = _encode32Chars[(int)(id >> 25) & 31];
-            charBuffer[8] = _encode32Chars[(int)(id >> 20) & 31];
-            charBuffer[9] = _encode32Chars[(int)(id >> 15) & 31];
-            charBuffer[10] = _encode32Chars[(int)(id >> 10) & 31];
-            charBuffer[11] = _encode32Chars[(int)(id >> 5) & 31];
-            charBuffer[12] = _encode32Chars[(int)id & 31];
+            FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref encodeChars);
```

Comment:
Would `FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref _encode32Chars[0])` produce less efficient code? It's fine the way it is, but I'm now paranoid that the local is somehow necessary for JIT to optimize away the bounds check.

Reply:
> that the local is somehow necessary for JIT to optimize away the bounds check

The bounds-check for `_encode32Chars[0]` won't be eliminated either way :sad:
Though I assume the JIT could to this on `static readonly`s. Don't know if this tracked anywhere.

I'll update it to the "inline" version w/o the local.